artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/RefCountMessage.java

```diff
@@ -49,6 +76,28 @@ public int getDurableCount() {
       return DURABLE_REF_COUNT_UPDATER.get(this);
    }
 
+   /** Sub classes may mark fired=true when they were explicitly called.
+    * E.g large message removed the file upon cancellation */
+   protected void markFired() {
+      fired = true;
+      if (firedAtomic != null) {
+         firedAtomic.set(true);
+      }
+   }
+
+   public RefCountMessage() {
+      if (DEBUG_REF_COUNT) {
+         AtomicBoolean refFired = this.firedAtomic;
+         List<Exception> crumbs = debugCrumbs;
+         crumbs.add(new Exception("new Instance"));
+         String clazz = this.getClass().getName();
+         // notice you can't have a direct reference on the LargeMessage Itself,
+         // Otherwise you would get a leak from itself
+         // RefCountMessageLeakTest is validating this scenario
+         CLEANER.register(this, () -> debug(clazz, refFired, crumbs));
```

Comment:
Even with the comment I'm not a huge fan of the lambda being used here because it is super easy for someone to later change this to capture 'this' inadvertently.  I'd recommend passing the state data to a static function that returns a Runnable just as an additional safety mechanism as that prevents a capture of 'this' in the Runnable.

Reply:
I actually amended this with a State Object.